src/System.IO.Pipelines.Networking.Tls/OpenSslConnectionContext.cs

```diff
@@ -0,0 +1,171 @@
+﻿using System.IO.Pipelines.Networking.Tls.Internal.OpenSsl;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines.Networking.Tls
+{
+    public class OpenSslConnectionContext : ISecureContext
+    {
+        private static readonly Task CachedTask = Task.FromResult(0);
+
+        private readonly OpenSslSecurityContext _securityContext;
+        private int _headerSize = 5; //5 is the minimum (1 for frame type, 2 for version, 2 for frame size)
+        private int _trailerSize = 16; //Used for MAC storage
+        private bool _readyToSend;
+        private IntPtr _ssl;
+        private readonly InteropBio.BioHandle _readBio;
+        private readonly InteropBio.BioHandle _writeBio;
+        private ApplicationProtocols.ProtocolIds _negotiatedProtocol;
+
+        public OpenSslConnectionContext(OpenSslSecurityContext securityContext, IntPtr ssl)
```

Comment:
Public types should not take IntPtr, but typed structures, preferably SafeHandles so the usage tracking isn't impossible.